pkg/beacon/relay/registry/storage.go

```diff
@@ -2,15 +2,21 @@ package registry
 
 import (
 	"fmt"
+	"strings"
 	"sync"
 
 	"github.com/keep-network/keep-common/pkg/persistence"
 
 	"encoding/hex"
 )
 
+const (
+	candidateMembershipPrefix = "candidate_membership"
+	approvedMembershipPrefix  = "approved_membership"
```

Comment:
I am thinking if we shouldn't use the [snapshot](https://github.com/keep-network/keep-ecdsa/blob/bd9800d7367aa665b745c7149e1040a3b16deaf0/pkg/registry/storage.go#L14) mechanism from ECDSA node. This is something our operators are familiar with and I think it makes sense here and in ECDSA:
- any key material produced by the node lands in `snapshot`
- when the group is approved, we copy from `snapshot` to `current` (this would be new)

We would need to "snapshot" the state anyway, and I think adding `candidate`/`approved` does not make sense if the snapshot is already there (this is essentially candidate).

Instead of passing the `GroupState` as a parameter, I would "snapshot" the signer as soon as the key material is produced. Then, once the group is approved on-chain, I would copy from the `snapshot` to the `current`. This way, we do not have to worry if the node has been restarted during the challenge period. Also, the "snapshot" would not store the group in the in-memory registry. This is exactly [how it is working](https://github.com/keep-network/keep-ecdsa/blob/bd9800d7367aa665b745c7149e1040a3b16deaf0/pkg/registry/keeps.go#L67-L73) in ECDSA.

Reply:
> This is something our operators are familiar with and I think it makes sense here and in ECDSA:
    - any key material produced by the node lands in snapshot
    - when the group is approved, we copy from snapshot to current (this would be new)

Well, this is not that simple. The snapshot mechanism is write-only. There is no logic in the code to read from it. This is a last-resort mechanism for manual restoration in case something bad happens. What ECDSA does is rather:

- any key material produced by the node lands in snapshot
- when the public key is published on-chain it is just normally stored in the regular storage (snapshot is untouched)

If the client shuts down between those steps, the key would not be published at all. That works well in ECDSA because the time between a snapshot and a regular save is quite short. In beacon v2, the time between result publication and approval (the challenge period) will be much longer so the client restart risk is much higher. If we use snapshot on result publication + store on approval, we don't have automatic restore after the client restarts out of the box. This is a problem because the client must be aware of what candidate groups he is a member of.
In the current approach, we would address this by adding an in-mem cache for candidate groups once we add the result approvals and challenges feature. If we want to go with snapshots, we'll need to make a huge refactor of the entire storage code. This can be done but then, I propose to do it in a separate PR.

Reply:
The solution I proposed requires refactoring of storage. The snapshot would still be read-only but we would need to be able to copy from the snapshot.

On a high level, the idea is not to load candidate members to the memory - even a separate cache - and instead copy the directory when a group is approved. I think this is simpler and has the advantage that copy is one disk write operation, we do not need to write in the loop which I think is less risky.

What I do not like in the current implementation is that looking at just `current` directory it is not possible to say which groups were approved and which were not. We are also mixing candidate key shares with approved key shares which is not the end of the world but I think we could make it cleaner.

If we want to treat snapshots as something special we write only one time and never touch them again, I think it's even better because by obeying this rule in the code, we avoid corrupting snapshot shares in case of a programming error. Looking at the storage level, we would have:

```
snapshot/
  0xAAA
  0xBBB
  0xCCC
current/
  candidate_0xAAA/
    membership_1
    membership_21
  0xBBB
    membership_33
    membership_55
  0xCCC
     membership_64
``` 

Then, when `0xAAA` gets approved, we `mv current/candidate_0xAAA current/0xAAA` and we finally land with:
```
snapshot/
  0xAAA
  0xBBB
  0xCCC
current/
  0xAAA/
    membership_1
    membership_21
  0xBBB
    membership_33
    membership_55
  0xCCC
     membership_64
```

Once the directory is renamed, we load all the members from `0xAAA` into the memory.

If we are fine with this layout, I would propose something like:

`groups.go`
```
// RegisterCandidateGroup registers that a candidate group with the given
// public key was successfully created.
func (g *Groups) RegisterCandidateGroup(
	signer *dkg.ThresholdSigner,
	groupId string,
) error {
	g.mutex.Lock()
	defer g.mutex.Unlock()

	groupPublicKey := groupKeyToString(signer.GroupPublicKeyBytes())

	membership := &Membership{
		Signer:      signer,
		ChannelName: groupId,
	}

	err := g.storage.saveCandidate(membership, groupID)
	if err != nil {
		return fmt.Errorf("could not persist membership to the storage: [%v]", err)
	}

	if groupState == GroupStateApproved {
		g.myGroups[groupPublicKey] = append(g.myGroups[groupPublicKey], membership)
	}

	return nil
}

// ApproveGroup registers that a candidate group with the given
// public key was successfully approved.
func (g *Groups) ApproveGroup(
	groupId string,
) error {
	err := g.storage.approveCandidate(groupId)
	if err != nil {
		return fmt.Errorf("could not promote group to approved in the storage: [%v]", err)
	}
     
        loadGroup(groupId)
}
```

`storage.go`
```
func (ps *persistentStorage) saveCandidate(
	membership *Membership,
	groupID string // use this as a directory name
) error {
  // (...)
}

func (ps *persistentStorage) approveCandidate(
	groupID string // use this as a directory name
) error {
  // (...)
}

func (ps *persistentStorage) read(groupID string) (<-chan *Membership, <-chan error) {
 // (...) load a single group
}

// Alternatively, we can combine approveCandidate(groupID string) with read(groupID string):
func (ps *persistentStorage) approveCandidate(
	groupID string // use this as a directory name
) (<-chan *Membership, <-chan error)  {
  // (...)
}
```

However, this is a bit of work and as you say, it's not the best idea to do it in this PR. If we agree with this approach, I'd clean up the code in this PR and prepare for the new version:
1. Get rid of `GroupState` and candidate/approved prefixes.
2. Rename `save` in `storage.go` to `saveCandidate`, accept `groupID string`, and prepend `candidate_` to group identifier when saving.
3. In `groups.go`, have only `RegisterCandidateGroup`.
4. Modify `storage.go` `readAll` to read candidate groups from disk. Make it a temporary change, until we refactor the storage. In a separate PR refactoring the storage, we should clean this up to read only approved groups and temporarily, approve the group right away, after it's saves as a candidate.

WDYT?

Reply:
See 641466e7

Reply:
Keeping this thread as unresolved so that we can refer it back when reworking the storage.